pkg/blockbuilder/tsdb.go

```diff
@@ -0,0 +1,398 @@
+					continue
+				}
+			} else {
+				// Copy the label set because both TSDB and the active series tracker may retain it.
+				copiedLabels = mimirpb.CopyLabels(nonCopiedLabels)
+				// Retain the reference in case there are multiple samples for the series.
+				if ref, err = app.Append(0, copiedLabels, s.TimestampMs, s.Value); err == nil {
+					continue
+				}
+			}
+			// TODO(v): not all errors should terminate; see how it ingester handles them
+			return false, err
+		}
+
+		for _, h := range ts.Histograms {
+			if h.Timestamp >= blockMax {
+				// We will process this sample in the next cycle.
+				allSamplesProcessed = false
+				continue
+			}
+			if recordProcessedBefore && h.Timestamp < lastBlockMax {
+				// This sample was already processed in the previous cycle.
+				continue
+			}
+			var (
+				ih *histogram.Histogram
+				fh *histogram.FloatHistogram
+			)
+
+			if h.IsFloatHistogram() {
+				fh = mimirpb.FromFloatHistogramProtoToFloatHistogram(&h)
+			} else {
+				ih = mimirpb.FromHistogramProtoToHistogram(&h)
+			}
+
+			if ref != 0 {
+				// If the cached reference exists, we try to use it.
+				if _, err = app.AppendHistogram(ref, copiedLabels, h.Timestamp, ih, fh); err == nil {
+					continue
+				}
+			} else {
+				// Copy the label set because both TSDB and the active series tracker may retain it.
+				copiedLabels = mimirpb.CopyLabels(nonCopiedLabels)
+				// Retain the reference in case there are multiple samples for the series.
+				if ref, err = app.AppendHistogram(0, copiedLabels, h.Timestamp, ih, fh); err == nil {
+					continue
+				}
+			}
+			// TODO(v): not all errors should terminate; see how it ingester handles them
+			return false, err
+		}
+
+		// Exemplars are not persisted in the block. So we skip them.
+	}
+
+	return allSamplesProcessed, app.Commit()
+}
+
+func (b *TSDBBuilder) getOrCreateTSDB(tenant tsdbTenant) (*userTSDB, error) {
+	b.tsdbsMu.RLock()
+	db := b.tsdbs[tenant]
+	b.tsdbsMu.RUnlock()
+	if db != nil {
+		return db, nil
+	}
+
+	b.tsdbsMu.Lock()
+	defer b.tsdbsMu.Unlock()
+
+	// Check again for DB in the event it was created in-between locks
+	var ok bool
+	db, ok = b.tsdbs[tenant]
+	if ok {
+		return db, nil
+	}
+
+	db, err := b.newTSDB(tenant)
+	if err != nil {
+		return nil, err
+	}
+
+	b.tsdbs[tenant] = db
+
+	return db, nil
+}
+
+func (b *TSDBBuilder) newTSDB(tenant tsdbTenant) (*userTSDB, error) {
+	udir := b.tsdbDir(tenant)
+	// Remove any previous TSDB dir. We don't need it.
+	if err := os.RemoveAll(udir); err != nil {
+		return nil, err
+	}
+	if err := os.MkdirAll(udir, os.ModePerm); err != nil {
+		return nil, err
+	}
+
+	userID := tenant.id
+	userLogger := util_log.WithUserID(userID, b.logger)
+
+	udb := &userTSDB{
+		userID: userID,
+	}
+
+	db, err := tsdb.Open(udir, userLogger, nil, &tsdb.Options{
+		RetentionDuration:           0,
+		MinBlockDuration:            2 * time.Hour.Milliseconds(),
+		MaxBlockDuration:            2 * time.Hour.Milliseconds(),
```

Comment:
The 1h is something controlled by the overarching block builder using config and processing 1h data appropriately from Kafka. The TSDB still needs to work with 2h block boundaries like rest of the Mimir in case block builder tries to process some samples before the 1h block we are looking at.